NetworkState: New API to generate changed state

cathay4t opened the pull request:
When using nmpolicy to modified current state and apply back, nmpolicy
generated a full state including modified and current state, this
causing NetworkManager persistent all network configuration to disk.
It become a overlap problem when the SRIOV of that interface is
controller by other component for example OpenShift networking-sriov
operator. Please refer to https://issues.redhat.com/browse/RHEL-15275
for full detail.

Introduced `NetworkState::gen_diff()` which generate new `NetworkState`
contains only the changed properties.

Changed nmpolicy to apply only changed state.

Changed `nmstatectl apply` to print out changed state instead of desired
state after state applied.

Unit test cases included.

liangwen12year commented:
There are some conflicts, please rebase it first.

cathay4t commented:
> There are some conflicts, please rebase it first.

All set

liangwen12year commented on rust/src/lib/unit_tests/gen_diff_test_files/change_dns_only/diff.yml:
```diff
@@ -0,0 +1,11 @@
+dns-resolver:
+  config:
+    search:
+    - example.com
```
I can not see the dns-search difference between the current state and the desired state with the example provided except that the DNS server order changed.

liangwen12year replied:
Aha, the `net_state.gen_diff()` just copies all the DNS settings over when the DNS settings are changed. Not sure if it would be better if we have `dns.gen_diff()`.

```
        if merged_state.dns.is_changed() {
            ret.dns = self.dns.clone();
        }
```

cathay4t replied:
This seems harmless. Let's fix that in follow up PR. Issue tracked at https://github.com/nmstate/nmstate/issues/2559

liangwen12year reviewed (commented):
Can you explain more in the commit message what is the reason prefer applying only the changed state instead of the desired state ?

cathay4t commented:
> Can you explain more in the commit message what is the reason prefer applying only the changed state instead of the desired state ?

I have updated the github top comment and git commit message for use case.

liangwen12year commented:
`test_policy_examples` unit test failed, please take a look at it.

cathay4t committed:
NetworkState: New API to generate changed state

When using nmpolicy to modified current state and apply back, nmpolicy
generated a full state including modified and current state, this
causing NetworkManager persistent all network configuration to disk.
It become a overlap problem when the SRIOV of that interface is
controller by other component for example OpenShift networking-sriov
operator. Please refer to https://issues.redhat.com/browse/RHEL-15275
for full detail.

Introduced `NetworkState::gen_diff()` which generate new `NetworkState`
contains only the changed properties.

Changed nmpolicy to apply only changed state.

Changed `nmstatectl apply` to print out changed state instead of desired
state after state applied.

Unit test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
> `test_policy_examples` unit test failed, please take a look at it.

Fixed. Thanks!
